### DISCUSSION TITLE
relative Link in next 13

### DISCUSSION THREAD

**Discussion by vgkhiem1405** [vercel/next.js#52016] [App Router]
Hello,
I'm new to using Next.js 13 and I have a question about the <Link> component.

I'm currently on the "**/users**" page and I want to navigate to "**/users/new**". 

Is there any way to use the <Link> component to go to "**/new**" relative to the current page? I don't want to use **<Link href="/users/new/"></Link>**.

Thanks.

**Comment by icyJoseph** [comment]
Have you ever managed to do this with "plain" HTML? 

I think, doing relative to the current path is tricky. You might think that `./new` or `new` should work, but those are still relative to the parent of `users`

**Reply by hesalx** [reply]
TL;DR Define you own URL resolution base with a slash at the end like `<base href="https://www.example.com/users/" />`

This is per the standard: [RFC 1808](https://datatracker.ietf.org/doc/html/rfc1808)

URL resolution is different from the common filesystem path resolution logic.
```js
> path.resolve("/users", "new")
'/users/new'
> path.resolve("/users", "/new")
'/new'
> path.resolve("/users", "./new")
'/users/new'
> path.resolve("/users", "../new")
'/new'
```

However, 
```js
> new URL("new", "https://example.com/users").href
'https://example.com/new'
> new URL("/new", "https://example.com/users").href
'https://example.com/new'
> new URL("./new", "https://example.com/users").href
'https://example.com/new'
> new URL("../new", "https://example.com/users").href
'https://example.com/new'
```

That is because, even if implicitly, URLs distinguish between a "directory" and a "document".
> Section 4. Step 6: The last segment of the base URL's path (anything following the rightmost slash "/", or the entire path if no slash is present) is removed and the embedded URL's path is appended in its place.

What it means is that URLs are relative to the current "directory". Which also means:
```
> new URL("new", "https://example.com/users/").href
'https://example.com/users/new'
> new URL("/new", "https://example.com/users/").href
'https://example.com/new'
> new URL("./new", "https://example.com/users/").href
'https://example.com/users/new'
> new URL("../new", "https://example.com/users/").href
'https://example.com/new'
```
Looks almost exactly (if you forget the slash) like path resolution, doesn't it?

But the slash at the end is frowned upon in the modern age. However, what the standard also says is that the URL resolution uses the "base URL". And that is something that can be defined by the developer!
```html
<head>
<base href="https://www.example.com/users/" />
</head>
<body>
<a href="new">I am https://www.example.com/users/new</a>
</body>
```

**Comment by hesalx** [comment]
@vgkhiem1405 As the are no details about your setup it isn't clear what approach suits you better but here are a few options:

1. Define a custom base URL in the `<head>` of the page with a slash at the end, e.g. 
```html
<base href="https://example.com/users/" />
```
This way `new` resolves to `/users/new`. How exactly you can construct the base URL varies greatly depending on whether use or want SSR/SSG, App Router, etc. As a hint, check out `useRouter` (if using Pages Router), `usePathname` (if using App Router), `pathname` from `getInitialProps` (SSG/SSR).

2. Resolve the URL yourself based on the current URL equipped with the knowledge that it's just a matter of an extra slash:
```jsx
const path = new URL("new", window.location.href + "/").href
return <a href={path}>new</a>
```
Again, depending on your setup you may want to use other methods of obtaining the current URL besides `window.location`.

You may prefer the first option if you don't want to know anything extra inside you component. However, you can always define your custom component that does the non-standard resolution under the hood.
```jsx
// does something like new URL(href, window.location.href + "/").href under the hood
<RelativeLink href={'new'}>New</RelativeLink>
```

**Comment by vgkhiem1405** [comment]
Hi @hesalx, @icyJoseph 
Thank you for your answers, but sorry for my question confusing you.
I'll say it again for clarity.

For example, I'm working on page **/users**
If I want to go to page **/users/new**, I'm using this code 
`<Link href="/users/new">create new</Link>`
you see that I already stay on **/users**. Using href="/users/new" is so tricky.
I have try use href="./new" or /new or....... but it always goes to page /new

So can i go to **/users/new** by another way?

**Reply by vgkhiem1405** [reply]
I wrote that code on **/users** page, not another page

**Reply by hesalx** [reply]
My answer is exactly about the case when you are already on the page `/users`, there was no confusion there. Code examples are just examples, I did not suggest to hardcode the entire URL to `/users` on another page. As I have mentioned, you either construct the `<base>` dynamically, so it always corresponds to the current page you are on, or you resolve the relative path (like `new`) based on the current page you are on. In both cases it assumes that the link to `new` is on the page `/users` **without explicitly mentioning** `/users` part in the URL.

If you find it inconvenient to write URL href explicitly every time, you should rather look at a more common approach, which is to define routes in a separate file and use it throughout the project.
```js
export const to = {
  Users: () => "/users",
  NewUser: () => "/users/new",
  ShowUser: ({ id }) => `/users/${id}`
}
```
```js
import { to } from '../routes'

export Component = () => {
  return <Link href={to.NewUser()}>create new</Link>
}
```
This is a very common pattern in React and Next.js apps.

**Reply by vgkhiem1405** [reply]
thank you very much @hesalx
